File: frontend/app/intellij-plugin/src/main/kotlin/io/mateu/ijp/api/SessionGuard.kt
package io.mateu.ijp.api

/**
 * Session-expiry re-auth (parity with the web's `onSessionExpired`). When a Mateu request comes back
 * 401 (the session token expired mid-session), the plugin gets ONE chance to re-authenticate before
 * the call fails: register a handler that re-auths (e.g. a modal prompt, a token refresh) and returns
 * whether the caller should retry. Synchronous by design — the IntelliJ HTTP client is blocking, and
 * a Swing re-auth dialog blocks anyway.
 *
 * Opt-in: with no handler registered, a 401 fails normally. The client retries at most once.
 */
object SessionGuard {

    @Volatile
    private var handler: (() -> Boolean)? = null

    /** Register (or clear, with null) the re-auth handler. Returns true to retry, false to give up. */
    fun onSessionExpired(h: (() -> Boolean)?) {
        handler = h
    }

    /** Invoked by the API client on a 401: true if re-authenticated and the caller should retry once. */
    fun handleSessionExpired(): Boolean =
        try {
            handler?.invoke() ?: false
        } catch (e: Exception) {
            false
        }
}

File: frontend/app/react-native/src/core/sessionGuard.ts
/**
 * Session-expiry re-auth (parity with the web's `onSessionExpired`, libs/mateu sessionGuard). When a
 * Mateu request comes back 401 (the Bearer token expired mid-session), the app gets ONE chance to
 * re-authenticate before the call fails. Register a handler that re-auths however the app does it
 * (refresh a token, prompt for credentials) and returns whether the caller should retry.
 *
 * Opt-in: with no handler registered, a 401 fails normally (no behaviour change). The client retries
 * the original request at most once.
 */
export type SessionExpiredHandler = () => boolean | Promise<boolean>;

let handler: SessionExpiredHandler | undefined;

/** Register (or clear, with undefined) the re-auth handler. */
export const onSessionExpired = (h: SessionExpiredHandler | undefined): void => {
  handler = h;
};

/**
 * Invoked by the API client on a 401. Returns true when the app re-authenticated and the caller
 * should retry once; false (no handler, the handler declined, or it threw) to fail normally.
 */
export const handleSessionExpired = async (): Promise<boolean> => {
  if (!handler) return false;
  try {
    return await handler();
  } catch {
    return false;
  }
};
